The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10008020175/job/27663891214) [(plain)](https://github.com/rust-lang/rust/commit/5791bbf5438fc0b5eb5755ccb5666dd557f6a273/checks/27663891214/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:692973e3d937129bcbf40652eb9f2f61becf3332)
Download action repository 'actions/upload-artifact@v4' (SHA:0b2256b8c012f0828dc542b3febcab082c67f72b)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 DONE 0.0s

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#13 0.425   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#13 0.439 Collecting boolean-py==4.0
#13 0.443   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#13 0.453 Collecting chardet==5.1.0
---
#13 3.791 Building wheels for collected packages: reuse
#13 3.792   Building wheel for reuse (pyproject.toml): started
#13 4.136   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 4.137   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#13 4.137   Stored in directory: /tmp/pip-ephem-wheel-cache-4f_fxwsi/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#13 4.140 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#13 4.163   Attempting uninstall: setuptools
#13 4.164     Found existing installation: setuptools 59.6.0
#13 4.165     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 4.165     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 4.165     Can't uninstall 'setuptools'. No files were found to uninstall.
#13 4.866 Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
#13 4.866 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 5.410 Collecting virtualenv
#13 5.448   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 5.525      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 76.6 MB/s eta 0:00:00
#13 5.564 Collecting distlib<1,>=0.3.7
#13 5.567   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 5.574      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 87.3 MB/s eta 0:00:00
#13 5.601 Collecting platformdirs<5,>=3.9.1
#13 5.605   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 5.638 Collecting filelock<4,>=3.12.2
#13 5.641   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 5.731 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.915 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 6.0s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      233408 kB
DirectMap2M:     6057984 kB
DirectMap1G:    12582912 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/8c3a94a1c79c67924558a4adf7fb6d98f5f0f741/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-8c3a94a1c79c67924558a4adf7fb6d98f5f0f741-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished `release` profile [optimized] target(s) in 29.60s
##[endgroup]
fmt check
##[error]Diff in /checkout/src/tools/tidy/src/style.rs at line 402:
         let is_test = file.components().any(|c| c.as_os_str() == "tests")
             || file.file_stem().unwrap() == "tests";
         let is_style = file.ends_with(style_file) || style_file.ends_with(file);
-        let is_style_test = is_test && file.parent().unwrap().ends_with(style_file.with_extension(""));
+        let is_style_test =
+            is_test && file.parent().unwrap().ends_with(style_file.with_extension(""));
         // scanning the whole file for multiple needles at once is more efficient than
         // executing lines times needles separate searches.
-        let any_problematic_line = !is_style && !is_style_test && problematic_regex.is_match(contents);
+        let any_problematic_line =
+            !is_style && !is_style_test && problematic_regex.is_match(contents);
         for (i, line) in contents.split('\n').enumerate() {
             if line.is_empty() {
                 if i == 0 {
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/tests/run-make/core-no-oom-handling/rmake.rs" "/checkout/src/tools/opt-dist/src/utils/artifact_size.rs" "/checkout/src/tools/opt-dist/src/utils/io.rs" "/checkout/src/tools/opt-dist/src/utils/mod.rs" "/checkout/src/tools/opt-dist/src/timer.rs" "/checkout/src/tools/opt-dist/src/metrics.rs" "/checkout/src/tools/opt-dist/src/main.rs" "/checkout/src/tools/opt-dist/src/bolt.rs" "/checkout/src/tools/opt-dist/src/tests.rs" "/checkout/src/tools/opt-dist/src/environment.rs" "/checkout/src/tools/opt-dist/src/exec.rs" "/checkout/tests/run-make/weird-output-filenames/rmake.rs" "/checkout/tests/run-make/weird-output-filenames/foo.rs" "/checkout/src/tools/rustdoc-gui-test/src/config.rs" "/checkout/src/tools/rustdoc-gui-test/src/main.rs" "/checkout/tests/run-make/lto-smoke-c/foo.rs" "/checkout/src/tools/miropt-test-tools/src/lib.rs" "/checkout/tests/run-make/print-to-output/rmake.rs" "/checkout/tests/run-make/lto-empty/rmake.rs" "/checkout/tests/run-make/lto-empty/lib.rs" "/checkout/tests/run-make/env-dep-info/rmake.rs" "/checkout/tests/run-make/env-dep-info/macro_use.rs" "/checkout/tests/run-make/env-dep-info/main.rs" "/checkout/tests/run-make/env-dep-info/macro_def.rs" "/checkout/tests/run-make/rust-lld/rmake.rs" "/checkout/tests/run-make/rust-lld/main.rs" "/checkout/src/tools/tidy/src/fluent_alphabetical.rs" "/checkout/src/tools/tidy/src/rustdoc_gui_tests.rs" "/checkout/src/tools/tidy/src/tests_revision_unpaired_stdout_stderr.rs" "/checkout/src/tools/tidy/src/unit_tests.rs" "/checkout/src/tools/tidy/src/run_make_tests.rs" "/checkout/src/tools/tidy/src/style/tests.rs" "/checkout/src/tools/tidy/src/unknown_revision.rs" "/checkout/src/tools/tidy/src/ext_tool_checks.rs" "/checkout/src/tools/tidy/src/features/version/tests.rs" "/checkout/src/tools/tidy/src/features/version.rs" "/checkout/src/tools/tidy/src/features/tests.rs" "/checkout/src/tools/tidy/src/tests_placement.rs" "/checkout/src/tools/tidy/src/pal.rs" "/checkout/src/tools/tidy/src/target_policy.rs" "/checkout/src/tools/tidy/src/target_specific_tests.rs" "/checkout/src/tools/tidy/src/deps.rs" "/checkout/src/tools/tidy/src/lib.rs" "/checkout/src/tools/tidy/src/extdeps.rs" "/checkout/src/tools/tidy/src/style.rs" "/checkout/tests/run-make/raw-dylib-inline-cross-dylib/rmake.rs" "/checkout/src/tools/tidy/src/unstable_book.rs" "/checkout/src/tools/tidy/src/alphabetical.rs" "/checkout/tests/run-make/raw-dylib-inline-cross-dylib/lib.rs" "/checkout/src/tools/tidy/src/bins.rs" "/checkout/tests/run-make/raw-dylib-inline-cross-dylib/lib_wrapper.rs" "/checkout/src/tools/tidy/src/main.rs" "/checkout/tests/run-make/raw-dylib-inline-cross-dylib/driver.rs" "/checkout/src/tools/tidy/src/edition.rs" "/checkout/src/tools/tidy/src/known_bug.rs" "/checkout/src/tools/tidy/src/fluent_period.rs" "/checkout/src/tools/tidy/src/fluent_used.rs" "/checkout/src/tools/tidy/src/features.rs" "/checkout/src/tools/tidy/src/iter_header.rs" "/checkout/src/tools/tidy/src/x_version.rs" "/checkout/src/tools/tidy/src/ui_tests.rs" "/checkout/src/tools/tidy/src/alphabetical/tests.rs" "/checkout/src/tools/tidy/src/error_codes.rs" "/checkout/src/tools/opt-dist/src/training.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Fri Jul 19 12:32:09 UTC 2024
  network time: Fri, 19 Jul 2024 12:32:09 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/src/tools/tidy/src/style.rs
+++ b/src/tools/tidy/src/style.rs
@@ -18,9 +18,9 @@
 // ignore-tidy-dbg
 
 use crate::walk::{filter_dirs, walk};
-use regex::RegexSet;
+use regex::RegexSetBuilder;
 use rustc_hash::FxHashMap;
-use std::{ffi::OsStr, path::Path};
+use std::{ffi::OsStr, path::Path, sync::LazyLock};
 
 #[cfg(test)]
 mod tests;
@@ -110,14 +110,24 @@ const ROOT_PROBLEMATIC_CONSTS: &[u32] = &[
     173390526, 721077,
 ];
 
+const LETTER_DIGIT: &[(char, char)] = &[('A', '4'), ('B', '8'), ('E', '3')];
+
 // Returns all permutations of problematic consts, over 2000 elements.
 fn generate_problematic_strings(
     consts: &[u32],
     letter_digit: &FxHashMap<char, char>,
 ) -> Vec<String> {
     generate_problems(consts, letter_digit)
-        .flat_map(|v| vec![v.to_string(), format!("{:x}", v), format!("{:X}", v)])
+        .flat_map(|v| vec![v.to_string(), format!("{:X}", v)])
         .collect()
+}
+
+static PROBLEMATIC_CONSTS_STRINGS: LazyLock<Vec<String>> = LazyLock::new(|| {
+    generate_problematic_strings(ROOT_PROBLEMATIC_CONSTS, &LETTER_DIGIT.iter().cloned().collect())
+});
+
+fn contains_problematic_const(trimmed: &str) -> bool {
+    PROBLEMATIC_CONSTS_STRINGS.iter().any(|s| trimmed.to_uppercase().contains(s))
 }
 
 const INTERNAL_COMPILER_DOCS_LINE: &str = "#### This error code is internal to the compiler and will not be emitted with normal Rust code.";
@@ -316,14 +326,14 @@ pub fn check(path: &Path, bad: &mut bool) {
         // We only check CSS files in rustdoc.
         path.extension().map_or(false, |e| e == "css") && !is_in(path, "src", "librustdoc")
     }
-    let problematic_consts_strings = generate_problematic_strings(
-        ROOT_PROBLEMATIC_CONSTS,
-        &[('A', '4'), ('B', '8'), ('E', '3')].iter().cloned().collect(),
-    );
+
     // This creates a RegexSet as regex contains performance optimizations to be able to deal with these over
     // 2000 needles efficiently. This runs over the entire source code, so performance matters.
-    let problematic_regex = RegexSet::new(problematic_consts_strings.as_slice()).unwrap();
-
+    let problematic_regex = RegexSetBuilder::new(PROBLEMATIC_CONSTS_STRINGS.as_slice())
+        .case_insensitive(true)
+        .build()
+        .unwrap();
+    let style_file = Path::new(file!());
     walk(path, skip, &mut |entry, contents| {
         let file = entry.path();
         let filename = file.file_name().unwrap().to_string_lossy();
@@ -389,10 +399,15 @@ pub fn check(path: &Path, bad: &mut bool) {
         let mut lines = 0;
         let mut last_safety_comment = false;
         let mut comment_block: Option<(usize, usize)> = None;
-        let is_test = file.components().any(|c| c.as_os_str() == "tests");
+        let is_test = file.components().any(|c| c.as_os_str() == "tests")
+            || file.file_stem().unwrap() == "tests";
+        let is_style = file.ends_with(style_file) || style_file.ends_with(file);
+        let is_style_test =
+            is_test && file.parent().unwrap().ends_with(style_file.with_extension(""));
         // scanning the whole file for multiple needles at once is more efficient than
         // executing lines times needles separate searches.
-        let any_problematic_line = problematic_regex.is_match(contents);
+        let any_problematic_line =
+            !is_style && !is_style_test && problematic_regex.is_match(contents);
         for (i, line) in contents.split('\n').enumerate() {
             if line.is_empty() {
                 if i == 0 {
@@ -451,7 +466,7 @@ pub fn check(path: &Path, bad: &mut bool) {
             if line.contains('\r') {
                 suppressible_tidy_err!(err, skip_cr, "CR character");
             }
-            if filename != "style.rs" {
+            if !is_style {
                 // Allow using TODO in diagnostic suggestions by marking the
                 // relevant line with `// ignore-tidy-todo`.
                 if trimmed.contains("TODO") && !trimmed.contains("ignore-tidy-todo") {
@@ -462,12 +477,8 @@ pub fn check(path: &Path, bad: &mut bool) {
                 if trimmed.contains("//") && trimmed.contains(" XXX") {
                     err("Instead of XXX use FIXME")
                 }
-                if any_problematic_line {
-                    for s in problematic_consts_strings.iter() {
-                        if trimmed.contains(s) {
-                            err("Don't use magic numbers that spell things (consider 0x12345678)");
-                        }
-                    }
+                if any_problematic_line && contains_problematic_const(trimmed) {
+                    err("Don't use magic numbers that spell things (consider 0x12345678)");
                 }
             }
             // for now we just check libcore

--- a/src/tools/tidy/src/style/tests.rs
+++ b/src/tools/tidy/src/style/tests.rs
@@ -1,17 +1,10 @@
 use super::*;
 
 #[test]
-fn test_generate_problematic_strings() {
-    let problematic_regex = RegexSet::new(
-        generate_problematic_strings(
-            ROOT_PROBLEMATIC_CONSTS,
-            &[('A', '4'), ('B', '8'), ('E', '3'), ('0', 'F')].iter().cloned().collect(), // use "futile" F intentionally
-        )
-        .as_slice(),
-    )
-    .unwrap();
-    assert!(problematic_regex.is_match("786357")); // check with no "decimal" hex digits - converted to integer
-    assert!(problematic_regex.is_match("589701")); // check with "decimal" replacements - converted to integer
-    assert!(problematic_regex.is_match("8FF85")); // check for hex display
-    assert!(!problematic_regex.is_match("1193046")); // check for non-matching value
+fn test_contains_problematic_const() {
+    assert!(contains_problematic_const("721077")); // check with no "decimal" hex digits - converted to integer
+    assert!(contains_problematic_const("524421")); // check with "decimal" replacements - converted to integer
+    assert!(contains_problematic_const(&(285 * 281).to_string())); // check for hex display
+    assert!(contains_problematic_const(&format!("{:x}B5", 2816))); // check for case-alternating hex display
+    assert!(!contains_problematic_const("1193046")); // check for non-matching value
 }

